---
navigation.icon: uil:bug
---

# Error handling

> Learn how to catch errors in different lifecycle.

さまざまなライフサイクルでエラーを検出する方法を学びます。

## Handling Errors

> Nuxt 3 is a full-stack framework, which means there are several sources of unpreventable user runtime errors that can happen in different contexts:

Nuxt 3 はフルスタック フレームワークです。つまり、さまざまなコンテキストで発生する可能性のある、防止できないユーザー ランタイム エラーの原因がいくつかあります。

> 1. Errors during the Vue rendering lifecycle (SSR + SPA)
> 1. Errors during API or Nitro server lifecycle
> 1. Server and client startup errors (SSR + SPA)
> 1. Errors downloading JS chunks

1. Vue レンダリング ライフサイクル中のエラー (SSR + SPA)
1. API または Nitro サーバーのライフサイクル中のエラー
1. サーバーおよびクライアントの起動エラー (SSR + SPA)
1. JS チャンクのダウンロード時のエラー

### Errors During the Vue Rendering Lifecycle (SSR + SPA)

> You can hook into Vue errors using [`onErrorCaptured`](https://vuejs.org/api/composition-api-lifecycle.html#onerrorcaptured).

[`onErrorCaptured`](https://vuejs.org/api/composition-api-lifecycle.html#onerrorcaptured) を使用して Vue エラーにフックできます。

> In addition, Nuxt provides a `vue:error` hook that will be called if any errors propagate up to the top level.

さらに、Nuxt は、エラーがトップレベルまで伝播した場合に呼び出される `vue:error` フックを提供します。

> If you are using an error reporting framework, you can provide a global handler through [`vueApp.config.errorHandler`](https://vuejs.org/api/application.html#app-config-errorhandler). It will receive all Vue errors, even if they are handled.

エラー報告フレームワークを使用している場合は、[`vueApp.config.errorHandler`](https://vuejs.org/api/application.html#app-config-errorhandler) を通じてグローバル ハンドラーを提供できます。 たとえ処理されたとしても、すべての Vue エラーを受け取ります。

#### Example With Global Error Reporting Framework 

グローバルエラー報告フレームワークの例

```js
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.config.errorHandler = (error, context) => {
    // ...
  }
})
```

### Server and Client Startup Errors (SSR + SPA)

サーバーおよびクライアントの起動エラー (SSR + SPA)

> Nuxt will call the `app:error` hook if there are any errors in starting your Nuxt application.

Nuxt アプリケーションの起動時にエラーが発生した場合、Nuxt は `app:error` フックを呼び出します。

This includes:

> * running Nuxt plugins
> * processing `app:created` and `app:beforeMount` hooks
> * rendering your Vue app to HTML (on the server)
> * mounting the app (on client-side), though you should handle this case with `onErrorCaptured` or with `vue:error`
> * processing the `app:mounted` hook

* Nuxtプラグインの実行
* `app:created` および `app:beforeMount` フックの処理
* Vue アプリを HTML にレンダリングします (サーバー上で)
* アプリを (クライアント側で) マウントします。ただし、このケースは `onErrorCaptured` または `vue:error` で処理する必要があります
* `app:mounted` フックの処理

### Errors During API or Nitro Server Lifecycle

API または Nitro サーバーのライフサイクル中のエラー

You cannot currently define a server-side handler for these errors, but can render an error page (see the next section).

現在、これらのエラーに対するサーバー側ハンドラーを定義することはできませんが、エラー ページを表示することはできます (次のセクションを参照)。

### Errors downloading JS chunks

JS チャンクのダウンロード中にエラーが発生しました

You might encounter chunk loading errors due to a network connectivity failure or a new deployment (which invalidates your old, hashed JS chunk URLs). Nuxt provides built-in support for handling chunk loading errors by performing a hard reload when a chunk fails to load during route navigation.

ネットワーク接続障害または新しいデプロイメント (古いハッシュ化された JS チャンク URL が無効になる) が原因で、チャンク読み込みエラーが発生する可能性があります。 Nuxt は、ルート ナビゲーション中にチャンクのロードに失敗した場合にハード リロードを実行することにより、チャンクのロード エラーを処理するための組み込みサポートを提供します。

You can change this behavior by setting `experimental.emitRouteChunkError` to `false` (to disable hooking into these errors at all) or to `manual` if you want to handle them yourself. If you want to handle chunk loading errors manually, you can check out the [the automatic implementation](https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/app/plugins/chunk-reload.client.ts) for ideas.

この動作を変更するには、`experimental.emitRouteChunkError` を `false` (これらのエラーへのフックをまったく無効にする) に設定するか、自分で処理したい場合は `manual` に設定します。 チャンク読み込みエラーを手動で処理したい場合は、[自動実装](https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/app/plugins/chunk-reload) をチェックアウトしてください。 .client.ts) のアイデアをお寄せください。

## Rendering an Error Page

エラーページのレンダリング

> When Nuxt encounters a fatal error (any unhandled error on the server, or an error created with `fatal: true` on the client) it will either render a JSON response (if requested with `Accept: application/json` header) or trigger a full-screen error page.

Nuxt が致命的なエラー (サーバー上の未処理のエラー、またはクライアント上の `fatal: true` で作成されたエラー) に遭遇すると、JSON 応答 (`Accept: application/json` ヘッダーでリクエストされた場合) をレンダリングするか、トリガーします。 全画面エラーページ。

> This error may occur during the server lifecycle when:

このエラーは、サーバーのライフサイクル中に次の場合に発生する可能性があります。

> * processing your Nuxt plugins
> * rendering your Vue app into HTML
> * a server API route throws an error

* Nuxt プラグインの処理
* Vue アプリを HTML にレンダリングする
* サーバー API ルートがエラーをスローする

> An error can also occur on the client side when:

次の場合にもクライアント側でエラーが発生する可能性があります。

> * processing your Nuxt plugins
> * before mounting the application (`app:beforeMount` hook)
> * mounting your app if the error was not handled with `onErrorCaptured` or `vue:error` hook
> * the Vue app is initialized and mounted in browser (`app:mounted`).

* Nuxt プラグインの処理
* アプリケーションをマウントする前 (`app:beforeMount` フック)
* エラーが `onErrorCaptured` または `vue:error` フックで処理されなかった場合にアプリをマウントする
* Vue アプリは初期化され、ブラウザーにマウントされます (`app:mounted`)。

> The lifecycle hooks are listed [here](/docs/api/advanced/hooks).

ライフサイクル フックは [ここ](/docs/api/advanced/hooks) にリストされています。

> You can customize this error page by adding `~/error.vue` in the source directory of your application, alongside `app.vue`. Although it is called an 'error page' it's not a route and shouldn't be placed in your `~/pages` directory. For the same reason, you shouldn't use `definePageMeta` within this page.

このエラー ページをカスタマイズするには、アプリケーションのソース ディレクトリに `app.vue` の横に `~/error.vue` を追加します。 これは「エラー ページ」と呼ばれていますが、ルートではないため、`~/pages` ディレクトリに配置すべきではありません。 同じ理由で、このページ内で `definePageMeta` を使用しないでください。

> The error page has a single prop - `error` which contains an error for you to handle.

 エラー ページには、処理すべきエラーを含む 1 つの小道具「error」があります。

> The `error` object provides the fields: `url`, `statusCode`, `statusMessage`, `message`, `description` and `data`. If you have an error with custom fields they will be lost; you should assign them to `data` instead. For custom errors we highly recommend to use `onErrorCaptured` composable that can be called in a page/component setup function or `vue:error` runtime nuxt hook that can be configured in a nuxt plugin.

`error` オブジェクトは、`url`、`statusCode`、`statusMessage`、`message`、`description`、および `data` のフィールドを提供します。 カスタム フィールドでエラーが発生した場合、カスタム フィールドは失われます。 代わりにそれらを「data」に割り当てる必要があります。 カスタム エラーの場合は、ページ/コンポーネントのセットアップ関数で呼び出すことができる `onErrorCaptured` コンポーザブル、または nuxt プラグインで設定できる `vue:error` ランタイム nuxt フックを使用することを強くお勧めします。

```ts
export default defineNuxtPlugin(nuxtApp => {
  nuxtApp.hook('vue:error', (err) => {
    //
  })
})
```

> When you are ready to remove the error page, you can call the `clearError` helper function, which takes an optional path to redirect to (for example, if you want to navigate to a 'safe' page).

エラー ページを削除する準備ができたら、「clearError」ヘルパー関数を呼び出すことができます。この関数は、リダイレクト先のオプションのパスを取得します (たとえば、「安全な」ページに移動する場合)。

> **Warning**
> Make sure to check before using anything dependent on Nuxt plugins, such as `$route` or `useRouter`, as if a plugin threw an error, then it won't be re-run until you clear the error.

`$route` や `useRouter` など、Nuxt プラグインに依存するものを使用する前に必ず確認してください。プラグインがエラーをスローした場合、エラーをクリアするまでプラグインは再実行されません。

> **Warning**
> If you are running on Node 16 and you set any cookies when rendering your error page, they will [overwrite cookies previously set](https://github.com/nuxt/nuxt/pull/20585). We recommend using a newer version of Node as Node 16 will reach end-of-life in September 2023.

Node 16 で実行していて、エラー ページをレンダリングするときに Cookie を設定すると、[以前に設定された Cookie が上書きされます](https://github.com/nuxt/nuxt/pull/20585)。 Node 16 は 2023 年 9 月にサポートが終了するため、新しいバージョンの Node を使用することをお勧めします。


### Example

```vue [error.vue]
<script setup lang="ts">
const props = defineProps({
  error: Object
})

const handleError = () => clearError({ redirect: '/' })
</script>

<template>
  <button @click="handleError">Clear errors</button>
</template>
```

## Error Helper Methods

### `useError`

* `function useError (): Ref<Error | { url, statusCode, statusMessage, message, description, data }>`

This function will return the global Nuxt error that is being handled.

> この関数は、処理されているグローバル Nuxt エラーを返します。

[ReadMore](/docs/api/composables/use-error)

### `createError`

* `function createError (err: { cause, data, message, name, stack, statusCode, statusMessage, fatal }): Error`

You can use this function to create an error object with additional metadata. It is usable in both the Vue and Nitro portions of your app, and is meant to be thrown.

この関数を使用すると、追加のメタデータを含むエラー オブジェクトを作成できます。 これはアプリの Vue 部分と Nitro 部分の両方で使用でき、スローされることを目的としています。

> If you throw an error created with `createError`:

`createError` で作成されたエラーをスローした場合:

> * on server-side, it will trigger a full-screen error page which you can clear with `clearError`.
> * on client-side, it will throw a non-fatal error for you to handle. If you need to trigger a full-screen error page, then you can do this by setting `fatal: true`.

* サーバー側では、全画面エラー ページがトリガーされますが、これは `clearError` でクリアできます。
* クライアント側では、致命的ではないエラーがスローされ、処理できます。 全画面エラー ページをトリガーする必要がある場合は、「fatal: true」を設定することでこれを行うことができます。

### Example

```vue [pages/movies/[slug\\].vue]
<script setup lang="ts">
const route = useRoute()
const { data } = await useFetch(`/api/movies/${route.params.slug}`)
if (!data.value) {
  throw createError({ statusCode: 404, statusMessage: 'Page Not Found' })
}
</script>
```

### `showError`

* `function showError (err: string | Error | { statusCode, statusMessage }): Error`

You can call this function at any point on client-side, or (on server side) directly within middleware, plugins or `setup()` functions. It will trigger a full-screen error page which you can clear with `clearError`.

> この関数は、クライアント側の任意の時点で呼び出すことも、(サーバー側で) ミドルウェア、プラグイン、または「setup()」関数内で直接呼び出すこともできます。 全画面エラー ページが表示されますが、これは `clearError` でクリアできます。

It is recommended instead to use `throw createError()`.

> 代わりに `throw createError()` を使用することをお勧めします。

[ReadMore](/docs/api/utils/show-error)

### `clearError`

* `function clearError (options?: { redirect?: string }): Promise<void>`

> This function will clear the currently handled Nuxt error. It also takes an optional path to redirect to (for example, if you want to navigate to a 'safe' page).

この関数は、現在処理されている Nuxt エラーをクリアします。 また、リダイレクト先のパスもオプションで指定します (たとえば、「安全な」ページに移動する場合)。

[ReadMore](/docs/api/utils/clear-error)

## Rendering Errors Within Your App

Rendering Errors Within Your App

> Nuxt also provides a `<NuxtErrorBoundary>` component that allows you to handle client-side errors within your app, without replacing your entire site with an error page.

Nuxt は、サイト全体をエラー ページで置き換えることなく、アプリ内でクライアント側のエラーを処理できるようにする `<NuxtErrorBoundary>` コンポーネントも提供します。

> This component is responsible for handling errors that occur within its default slot. On client-side, it will prevent the error from bubbling up to the top level, and will render the `#error` slot instead.

このコンポーネントは、デフォルトのスロット内で発生したエラーを処理する責任があります。 クライアント側では、エラーが最上位レベルまでバブルするのを防ぎ、代わりに `#error` スロットをレンダリングします。

> The `#error` slot will receive `error` as a prop. (If you set `error = null` it will trigger re-rendering the default slot; you'll need to ensure that the error is fully resolved first or the error slot will just be rendered a second time.)

`#error` スロットは prop として `error` を受け取ります。 (「error = null」を設定すると、デフォルト スロットの再レンダリングがトリガーされます。最初にエラーが完全に解決されていることを確認する必要があります。そうしないと、エラー スロットが 2 回目にレンダリングされるだけになります。)

> **Note**  
> If you navigate to another route, the error will be cleared automatically.

別のルートに移動すると、エラーは自動的に解除されます。

### Example

```vue [pages/index.vue]
<template>
  <!-- some content -->
  <NuxtErrorBoundary @error="someErrorLogger">
    <!-- You use the default slot to render your content -->
    <template #error="{ error, clearError }">
      You can display the error locally here: {{ error }}
      <button @click="clearError">
        This will clear the error.
      </button>
    </template>
  </NuxtErrorBoundary>
</template>
```

[LinkExample](/docs/examples/advanced/error-handling)
